### PR TITLE
Item reviews in React

### PR DESCRIPTION
**Pull request by bhollis**
This is an in-progress branch to move the item reviews components to React. The data model needs to switch away from mutating in place for this to work. @48klocs you can take it from here.

Note that I removed the ratings histogram since it was D1 only.

### CODE REVIEW

**Comment by bhollis** [comment]
You may find the immer library to be helpful: https://github.com/mweststrate/immer

File: src/app/destinyTrackerApi/d2-reviewDataCache.ts [hunk 1]
```diff
@@ -140,13 +151,17 @@ class D2ReviewDataCache {
...
-        const matchingScore = this._getMatchingItemByReviewKey(bulkRanking);
+        const cachedItem = this._getMatchingItemByReviewKey(bulkRanking);
+
+        if (cachedItem) {
+          const updatedCachedItem = produce(cachedItem, (newCachedItem) => {
+            newCachedItem.fetchResponse = bulkRanking;
+            newCachedItem.lastUpdated = new Date();
+            newCachedItem.overallScore = this._getScore(bulkRanking);
+            newCachedItem.ratingCount = bulkRanking.votes.total;
+          });
```
Comment (bhollis) [inline src/app/destinyTrackerApi/d2-reviewDataCache.ts]: You don't need immer for this - it's equivalent to:

```js
const updatedCachedItem = {
  ...cachedItem.
  fetchResponse: bulkRanking,
  lastUpdated: new Date(),
  overallScore: this._getScore(bulkRanking),
  ratingCount = votes.total
};
```

Immer really shines when you have deeply nested objects that you want to mess with.

File: src/app/destinyTrackerApi/d2-reviewDataCache.ts [hunk 2]
```diff
@@ -31,6 +32,16 @@ class D2ReviewDataCache {
...
   }
 
+  _replaceRatingData(oldRatingData: D2RatingData, newRatingData: D2RatingData) {
+    const index = this._itemStores.indexOf(oldRatingData);
+
+    if (index < 0) {
+      return;
+    }
+
+    this._itemStores[index] = newRatingData;
```
Comment (bhollis) [inline src/app/destinyTrackerApi/d2-reviewDataCache.ts]: This is still a mutating update. You need to create an entirely new array if you mean for them to be completely updated. You may want to switch the store to be keyed off of something else though.

**Comment by bhollis** [comment]
OK, I think this is done. @48klocs could you take a look over it?

I ended up separating the Redux store for reviews/ratings into three pieces:

1. `ratings` - the summary ratings that we load in bulk for items.
2. `reviews` - detailed review information for items, which we load one at a time.
3. `userReviews` - in-progress/cached user reviews.

By splitting them up I can select out just what I need, and avoid updating things in general.

Now that all the state is in Redux, I got rid of all the classes in favor of just functions. I also introduced `redux-thunk` so we don't have to directly reference `store` and can use provided state and dispatch when updating.

Overall everything should work the same, it's just all driven off Redux. We can do fun stuff like caching reviews much easier now.

**Comment by 48klocs** [comment]
Nice - one of the missing pieces on wish lists is updating the store as new items get added to it, but this should lay down the pattern to patch up that hole.

**Pull request merged**